File: frontend/src/lib/pricing.ts
import type { Part } from "@/lib/api";

export interface LinePricing {
  bomQty: number;
  minQty: number;
  batchCost: number | null;
  unitCost: number | null;
  unitCostOverride: boolean;
  orderQty: number;
  packsOrdered: number;
  lineTotal: number | null;
  unitCostExpression: string;
  batchNote: string;
  needsBatchOrder: boolean;
}

function roundQty(value: number): number {
  return Math.abs(value - Math.round(value)) < 1e-9
    ? Math.round(value)
    : value;
}

export function computeLinePricing(part: Part): LinePricing {
  const bomQty = Math.max(part.quantity || 0, 0);
  const minQty = Math.max(part.price_min_qty || 1, 1);
  const batchCost =
    part.price_batch_cost != null && !Number.isNaN(part.price_batch_cost)
      ? part.price_batch_cost
      : null;
  const unitOverride = part.unit_cost != null && !Number.isNaN(part.unit_cost);

  let unitCost: number | null;
  let unitCostExpression: string;

  if (unitOverride) {
    unitCost = part.unit_cost!;
    unitCostExpression = `manual: ${formatCurrency(unitCost)}/ea`;
  } else if (batchCost != null && minQty > 0) {
    unitCost = batchCost / minQty;
    unitCostExpression = `${formatCurrency(batchCost)} ÷ ${roundQty(minQty)} = ${formatCurrency(unitCost)}/ea`;
  } else {
    unitCost = null;
    unitCostExpression = "";
  }

  let packsOrdered = 1;
  let orderQty = bomQty;
  let lineTotal: number | null = null;
  let batchNote = "";

  if (minQty > 1) {
    packsOrdered = bomQty > 0 ? Math.max(1, Math.ceil(bomQty / minQty)) : 0;
    orderQty = packsOrdered * minQty;
    if (batchCost != null) {
      lineTotal = bomQty > 0 ? packsOrdered * batchCost : 0;
      batchNote = `${packsOrdered}× pack of ${roundQty(minQty)} @ ${formatCurrency(batchCost)}`;
    } else if (unitCost != null) {
      lineTotal = bomQty > 0 ? orderQty * unitCost : 0;
      batchNote = `order ${roundQty(orderQty)} (min pack ${roundQty(minQty)})`;
    } else {
      batchNote = `min pack ${roundQty(minQty)}`;
    }
  } else if (unitCost != null && bomQty > 0) {
    lineTotal = bomQty * unitCost;
  }

  const needsBatchOrder = bomQty > 0 && orderQty > bomQty;

  return {
    bomQty,
    minQty,
    batchCost,
    unitCost,
    unitCostOverride: unitOverride,
    orderQty,
    packsOrdered,
    lineTotal,
    unitCostExpression,
    batchNote,
    needsBatchOrder,
  };
}

export function summarizeProjectPricing(parts: Part[]): {
  lineCount: number;
  pricedLines: number;
  projectTotal: number;
} {
  const lines = parts.map(computeLinePricing);
  const priced = lines.filter((line) => line.lineTotal != null);
  return {
    lineCount: parts.length,
    pricedLines: priced.length,
    projectTotal: priced.reduce((sum, line) => sum + (line.lineTotal ?? 0), 0),
  };
}

export function formatCurrency(value: number | null | undefined): string {
  if (value == null || Number.isNaN(value)) return "—";
  return new Intl.NumberFormat("en-US", {
    style: "currency",
    currency: "USD",
    minimumFractionDigits: 2,
    maximumFractionDigits: 4,
  }).format(value);
}

export function parseOptionalNumber(raw: string): number | null {
  const trimmed = raw.trim();
  if (!trimmed) return null;
  const value = Number(trimmed);
  return Number.isFinite(value) && value >= 0 ? value : null;
}

File: backend/services/pricing.py
"""McMaster-style batch / pack pricing for BOM line totals."""

from __future__ import annotations

import math
from dataclasses import dataclass

from backend.models.part import Part


@dataclass(frozen=True)
class LinePricing:
    bom_qty: float
    min_qty: float
    batch_cost: float | None
    unit_cost: float | None
    unit_cost_override: bool
    order_qty: float
    packs_ordered: int
    line_total: float | None
    unit_cost_expression: str
    batch_note: str


def _round_qty(value: float) -> float:
    if math.isclose(value, round(value), rel_tol=0, abs_tol=1e-9):
        return float(round(value))
    return value


def compute_line_pricing(part: Part) -> LinePricing:
    """
    Derive unit and line totals from BOM quantity and batch fields.

    When ``price_min_qty`` > 1, McMaster often sells in packs — order quantity
    rounds up to whole packs and line total uses ``price_batch_cost`` per pack.
    """
    bom_qty = max(float(part.quantity or 0), 0.0)
    min_qty = max(float(part.price_min_qty or 1), 1.0)
    batch_cost = part.price_batch_cost
    unit_override = part.unit_cost is not None

    if unit_override:
        unit_cost = float(part.unit_cost)
        unit_expression = f"manual: ${unit_cost:.4f}/ea"
    elif batch_cost is not None and min_qty > 0:
        unit_cost = float(batch_cost) / min_qty
        unit_expression = f"${float(batch_cost):.2f} ÷ {_round_qty(min_qty)} = ${unit_cost:.4f}/ea"
    else:
        unit_cost = None
        unit_expression = ""

    packs = 1
    batch_note = ""
    if min_qty > 1:
        packs = max(1, math.ceil(bom_qty / min_qty)) if bom_qty > 0 else 0
        order_qty = packs * min_qty
        if batch_cost is not None:
            line_total = packs * float(batch_cost) if bom_qty > 0 else 0.0
            batch_note = (
                f"{packs}× pack of {_round_qty(min_qty)} @ ${float(batch_cost):.2f}"
            )
        elif unit_cost is not None:
            line_total = order_qty * unit_cost if bom_qty > 0 else 0.0
            batch_note = f"order {_round_qty(order_qty)} (min pack {_round_qty(min_qty)})"
        else:
            line_total = None
            batch_note = f"min pack {_round_qty(min_qty)}"
    else:
        order_qty = bom_qty
        if unit_cost is not None and bom_qty > 0:
            line_total = bom_qty * unit_cost
        else:
            line_total = None

    if bom_qty > 0 and order_qty > bom_qty and not batch_note:
        batch_note = f"order {_round_qty(order_qty)} to meet minimum"

    return LinePricing(
        bom_qty=bom_qty,
        min_qty=min_qty,
        batch_cost=batch_cost,
        unit_cost=unit_cost,
        unit_cost_override=unit_override,
        order_qty=order_qty,
        packs_ordered=packs,
        line_total=line_total,
        unit_cost_expression=unit_expression,
        batch_note=batch_note,
    )


def summarize_project_pricing(parts: list[Part]) -> dict[str, float | int]:
    totals = [compute_line_pricing(part).line_total for part in parts]
    priced = [value for value in totals if value is not None]
    return {
        "line_count": len(parts),
        "priced_lines": len(priced),
        "project_total": sum(priced) if priced else 0.0,
    }
